streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.utils.Bytes;
+
+/**
+ * ProcessorMetadata to be access and populated by processor node. This will be committed along with
+ * offset
+ */
+public class ProcessorMetadata {
+
+    // Does this need to be thread safe? I think not since there's one per task
+    private final Map<Bytes, byte[]> globalMetadata;
```

Comment:
Not sure if I can follow? Why is this a key-value pair? And why are the types `<Bytes, byte[]>`?

The `OffsetAndMetadata` interface takes a `String metadata` argument. That is why we encode the current `streamTime` we store as metadata as `String` encoded using base64.

If we assume that there might be multiple processors inside a task, I understand that we need a mapping from processor name to metadata, but the map should be `<String,String>` (or maybe `<String,Long>`)?

And we would encode it as concatenation of `<streamTime><numberOfEntries><key,value><key,value>` with key being the processor name, and value being the encoded timestamp (if we use `Long` in the map, we convert the long to base64 internally?

Reply:
> Why is this a key-value pair

My original intention is to make it flexible so that even different processor can commit/store multiple key/value pairs. So the key/value could be anything the processor choose. Now I think maybe we could also add a namespace to prevent collision of key from different processor. Namespace can be processor name.

> And why are the types <Bytes, byte[]>

Why not `<String, Long>`? I think `Long` is not flexible enough. For emit final use case, `Long` is fine. 
Why not `<String, String>`? I think it's more steps for processor to serialize it to `String` in some cases. e.g. For `Long`, processor needs to serialize it to `byte[]` and then `String`? So why not pass `byte[]` directly and serialize them finally to `String` for `OffsetAndMetadata`?

Reply:
Might be pre-mature optimization to make it generic. I would stick to just what we need. That is also why I think `<String,Long>` might just be good enough as we intend to only use it for "emit final".

> So why not pass byte[] directly and serialize them finally to String for OffsetAndMetadata?

My point is, that we should avoid to change the format/serialization twice, but only once. Either we accept `long` and let the runtime do `long -> String` directly. Or we just use `String` and the runtime has nothing to do but to concatenate string, and the processor needs to do the `long -> String` conversion. In both cases it's a single translation step. Introducing `byte[]` add a second translation step: what would be the advantage of having two steps instead of one?